recipes/LibriSpeech/ASR/CTC/train_with_wav2vec_k2.py

```diff
@@ -0,0 +1,557 @@
+        if hasattr(self.hparams, "upsampling") and self.hparams.upsampling:
+            logits = logits.view(
+                logits.shape[0], -1, self.hparams.output_neurons
+            )
+
+        p_ctc = self.hparams.log_softmax(logits)
+        return p_ctc, wav_lens
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC+NLL) given predictions and targets."""
+
+        p_ctc, wav_lens = predictions
+
+        ids = batch.id
+
+        # Sort batch to be descending by length of wav files, which is demanded by k2
+        if self.hparams.sorting == "ascending":
+            p_ctc = torch.flip(p_ctc, (0,))
+            wav_lens = torch.flip(wav_lens, (0,))
+            texts = [batch.wrd[i] for i in reversed(range(len(batch.wrd)))]
+            ids = [ids[i] for i in reversed(range(len(ids)))]
+        elif self.hparams.sorting == "descending":
+            texts = batch.wrd
+        elif self.hparams["sorting"] == "random":
+            indices = torch.argsort(wav_lens, descending=True)
+            p_ctc = p_ctc[indices]
+            wav_lens = wav_lens[indices]
+            texts = [batch.wrd[i] for i in indices]
+            ids = [ids[i] for i in indices]
+        else:
+            raise NotImplementedError(
+                "Only random, ascending or descending sorting is "
+                "implemented, but got {}".format(self.hparams.sorting)
+            )
+
+        is_training = stage == sb.Stage.TRAIN
+        loss = self.hparams.ctc_cost(
+            log_probs=p_ctc,
+            input_lens=wav_lens,
+            graph_compiler=self.graph_compiler,
+            texts=texts,
+            is_training=is_training,
+        )
+
+        if stage == sb.Stage.VALID:
+            # Decode token terms to words
+            lattice = sbk2.lattice_decoder.get_lattice(
+                p_ctc,
+                wav_lens,
+                self.decoder["decoding_graph"],
+                search_beam=self.hparams.test_search_beam,
+                output_beam=self.hparams.test_output_beam,
+                ac_scale=self.hparams.ac_scale,
+                max_active_states=self.hparams.test_max_active_state,
+                min_active_states=self.hparams.test_min_active_state,
+            )
+            # 1best decoding for fast valid
+            paths = {"1best": sbk2.lattice_decoder.one_best_decoding(lattice)}
+
+        if stage == sb.Stage.TEST:
+            lattice = sbk2.lattice_decoder.get_lattice(
+                p_ctc,
+                wav_lens,
+                self.decoder["decoding_graph"],
+                search_beam=self.hparams.test_search_beam,
+                output_beam=self.hparams.test_output_beam,
+                ac_scale=self.hparams.ac_scale,
+                max_active_states=self.hparams.test_max_active_state,
+                min_active_states=self.hparams.test_min_active_state,
+            )
+            paths = self.decoder["decoding_method"](lattice)
```

Comment:
usually the decoding is performed inside of the `compute_forward`, could you please move this block there?

Reply:
I disagree; computing forward has nothing to do with lattice decoding.
First, there can be ways to decode a forward's results (onebest, n-gram-rescoring, RNN rescoring).
Second, during training, `compute_forward` should perform any decoding.
Third, to extract ASR based phone posteriorgram (PPG) from the model, no ASR decoding is needed.

Reply:
I see that the other recipes returns a p_tokens from the `compute_forward`, I'll stay compliant with the other recipes and make the change.

Reply:
DONE